Destructuring

JonasWanke opened the pull request:
<!-- Please enter the corresponding issue ID: -->


<!-- Add the breaking label (PR: BREAKING) if applicable. -->

<!-- Please summarize your changes: -->
![image](https://user-images.githubusercontent.com/19330937/206871610-f79abf49-1504-46b4-b6cf-e691544a939e.png)

The duplicate `#` in hints and their order are not looking good, but I want to try using LSP inline values for them instead: https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_inlineValue.

This PR also fixes the semantic tokens for identifiers by passing down the context-inferred type and only marking them as functions when they are actually called.


### Checklist
<!-- Please check if your PR fulfills the following requirements: -->

- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

JonasWanke committed:
Add patterns (RCST, CST, AST, HIR)

JonasWanke committed:
Merge branch 'iterable' into patterns

JonasWanke committed:
Fix errors after merge

JonasWanke committed:
Add builtin toDebugText

JonasWanke committed:
Fix assignment parser test

JonasWanke committed:
Fix destructuring parsing

JonasWanke committed:
Ignore irrelevant Clippy complaint

JonasWanke committed:
Fix Clippy complaints and other minor changes

JonasWanke committed:
Lower destructuring assignment to MIR

JonasWanke committed:
Add missing trailing commas

JonasWanke committed:
Map HIR pattern identifiers to AST identifiers i/o AST strings

JonasWanke committed:
Fix destructuring value hints

JonasWanke committed:
Support struct destructuring

JonasWanke committed:
Handle invalid expressions in patterns

JonasWanke committed:
Add ast::LoweringContext::create_error(…)

JonasWanke committed:
Add long-term TODO for pattern panic messages

JonasWanke committed:
Avoid duplicate visits in LSP references

JonasWanke committed:
Remove invalid test case

JonasWanke committed:
Support struct shorthands

JonasWanke committed:
Fix semantic tokens for identifiers

MarcelGarus commented:
While meeting yesterday, we also decided that the Candy code should be updated to use pattern matching where appropriate.

JonasWanke committed:
Support struct brackets in VS Code language configuration

JonasWanke committed:
Fix indentation rules for VS Code

Add one for text in double quotes and remove special casing for text in parentheses / brackets / curly braces.

JonasWanke committed:
Fix module path parsing

JonasWanke committed:
Implement Display for PatternIdentifierId

JonasWanke committed:
Use display formatting for compiler errors in CLI

JonasWanke committed:
Keep child for errors in HIR patterns

JonasWanke committed:
Generate debug files next to actual file

JonasWanke committed:
Show module for compiler errors in CLI

JonasWanke committed:
Implement Display for CompilerError

JonasWanke committed:
Don't parse assignment in pattern

Otherwise, incomplete code like `[foo, bar = abc` (with a missing `]`) would parse the whole line as a pattern instead of closing the pattern before `=` and parsing the right-hand side normally.

JonasWanke committed:
Don't parse colon in struct field if it's part of an assignment

JonasWanke committed:
Add missing trailing commas

JonasWanke committed:
Shorten unknown reference error message

JonasWanke committed:
Use destructuring in Core

JonasWanke committed:
Add short-term TODOs

JonasWanke committed:
Merge branch 'main' into patterns

JonasWanke committed:
Fix errors after merge

JonasWanke committed:
Build Core in CI

JonasWanke committed:
Fix outdated import

MarcelGarus commented on packages/Core/controlFlow.candy:
```diff
@@ -1,7 +1,7 @@
+[equals] = use "..equality"
 function = use "..function"
-isSymbol = (use "..symbol").is
+symbol = use "..symbol"
```
I was confused why this didn’t cause code below to change. Turns out, the symbol module is never used. So I guess you can also remove all of this

JonasWanke committed:
Remove unused import

JonasWanke committed:
Format Markdown files

JonasWanke committed:
Specify missing languages in Markdown code blocks

JonasWanke committed:
Add documentation for lists

JonasWanke committed:
Document struct shorthands

JonasWanke committed:
Document destructuring

JonasWanke committed:
Fix language documentation formatting

JonasWanke committed:
Fix plural in list destructuring error message

MarcelGarus commented on language.md:
```diff
@@ -134,6 +137,20 @@ Green
 
+### Lists
+
+A list is a compound object that holds aa collection of elements identified by integer index.
```
```suggestion
A list is a compound object that holds a collection of elements identified by integer index.
```

JonasWanke committed:
Fix typo

Co-authored-by: Marcel Garus <marcel.garus@gmail.com>
